[NUTCH-2856] Implement a protocol-smb plugin based on hierynomus/smbj

HiranChaudhuri opened the pull request:
[NUTCH-2856] Implement a protocol-smb plugin based on hierynomus/smbj

Draft version of a protocol-smb plugin. Lots of todo comments still,
but it seems to work.

Commit:
[NUTCH-2856] Implement a protocol-smb plugin based on hierynomus/smbj

Draft version of a protocol-smb plugin. Lots of todo comments still,
but it seems to work.

lewismc commented:
Moving this to DRAFT status and acknowledging the PR @HiranChaudhuri thank you.
I will try to perform a full review this coming week... thank you.

lewismc commented on src/plugin/protocol-smb/ivy.xml:
```diff
@@ -0,0 +1,47 @@
+    <dependency org="net.engio" name="mbassador" rev="1.3.0"/>
+    <dependency org="org.bouncycastle" name="bcprov-jdk18on" rev="1.75"/>
+    <dependency org="com.hierynomus" name="asn-one" rev="0.6.0"/>
+    <dependency org="commons-io" name="commons-io" rev="2.17.0"/>
```
`commons-io` is already defined in `ivy/ivy.xml` (albeit an older version). It looks like you are only using this dependency for the `IOUtils.toByteArray` method so I wonder we can rework that and remove the dependency?
What about the methods in [DeflateUtils.java](https://github.com/apache/nutch/blob/master/src/java/org/apache/nutch/util/DeflateUtils.java). Would any of those work?

lewismc commented on src/plugin/protocol-smb/src/java/org/apache/nutch/protocol/smb/Smb.java:
```diff
@@ -0,0 +1,292 @@
+                ProtocolStatus.STATUS_SUCCESS
+              );
+            } else if (share.fileExists(path)) {
+              // todo: how can we store this, and maybe more metadata?
```
What are you thinking could be added?

lewismc commented on src/plugin/protocol-smb/src/java/org/apache/nutch/protocol/smb/Smb.java:
```diff
@@ -0,0 +1,292 @@
+      throw new IllegalArgumentException("Config parameter 'smb.password' not set.");
+    }
+    domain = conf.getTrimmed("smb.domain");
+    contentLimit = conf.getInt("smb.content.limit", Integer.MAX_VALUE);
```
Please take a look at [http.content.limit](https://github.com/apache/nutch/blob/master/conf/nutch-default.xml#L255) and [file.content.limit](https://github.com/apache/nutch/blob/master/conf/nutch-default.xml#L511)... I would suggest we mimic those values for now.

lewismc commented on src/plugin/protocol-smb/src/java/org/apache/nutch/protocol/smb/Smb.java:
```diff
@@ -0,0 +1,292 @@
+              // todo: we run into issues if the file is bigger than 2 GB. I made the limit configurable
+              // but e.g. zip can no longer be evaluated if too big.
+              if (fetchSize > contentLimit) {
+                LOG.warn("trunkating {}", urlstr);
```
Cleanup logging + spelling

lewismc commented on src/plugin/protocol-smb/src/java/org/apache/nutch/protocol/smb/Smb.java:
```diff
@@ -0,0 +1,292 @@
+              long fetchSize = fileSize;
+              metadata.add("fileSize", String.valueOf(fileSize));
+
+              // todo: we run into issues if the file is bigger than 2 GB. I made the limit configurable
```
Why do we run into issues? Does the logic reside in the SMBClient?

lewismc commented on src/plugin/protocol-smb/src/java/org/apache/nutch/protocol/smb/Smb.java:
```diff
@@ -0,0 +1,292 @@
+                metadata.add("truncated", String.valueOf(fetchSize));
+              }
+
+              bytes = IOUtils.toByteArray(fileIn, fetchSize); // read inputstream into byte array
```
As I mentioned this one line is the only usage of `commons-io`, I think we can just use this from the `ivy/ivy.xml`.

lewismc commented on src/plugin/protocol-smb/src/java/org/apache/nutch/protocol/smb/Smb.java:
```diff
@@ -0,0 +1,292 @@
+
+              bytes = IOUtils.toByteArray(fileIn, fetchSize); // read inputstream into byte array
+
+              LOG.warn("retrieved {} bytes", bytes.length);
```
Cleanup logging

lewismc commented on src/plugin/protocol-smb/src/java/org/apache/nutch/protocol/smb/Smb.java:
```diff
@@ -0,0 +1,292 @@
+                int b = bytes[i] & 0xFF;
+                sb.append(" ").append(HEX_ARRAY[b>>>4]).append(HEX_ARRAY[b & 0xF]);
+              }
+              LOG.warn("retrieved {} bytes starting with {}", bytes.length, sb.toString());
```
Cleanup logging

lewismc commented on src/plugin/protocol-smb/src/java/org/apache/nutch/protocol/smb/Smb.java:
```diff
@@ -0,0 +1,292 @@
+                sb.append(" ").append(HEX_ARRAY[b>>>4]).append(HEX_ARRAY[b & 0xF]);
+              }
+              LOG.warn("retrieved {} bytes starting with {}", bytes.length, sb.toString());
+              LOG.warn("metadata={}", metadata);
```
Cleanup logging

lewismc commented on src/plugin/protocol-smb/src/java/org/apache/nutch/protocol/smb/Smb.java:
```diff
@@ -0,0 +1,292 @@
+            } else {
+              // communicate error
+              String message = "File not found: " + urlstr;
+              LOG.warn(message);
```
Cleanup logging

lewismc commented on src/plugin/protocol-smb/src/java/org/apache/nutch/protocol/smb/Smb.java:
```diff
@@ -0,0 +1,292 @@
+
+        throw new UnsupportedOperationException("neither directory nor file: " + urlstr);
+    } catch(Exception e) {
+      LOG.error("Could not get protocol output for " + urlstr, e);
```
Use parameterized logging

lewismc commented on src/plugin/protocol-smb/src/java/org/apache/nutch/protocol/smb/Smb.java:
```diff
@@ -0,0 +1,292 @@
+  @Override
+  public BaseRobotRules getRobotRules(Text url, CrawlDatum datum,
+      List<Content> robotsTxtContent) {
+    LOG.debug("getRobotRules({}, {}, {})", url, datum, robotsTxtContent);
```
Is this necessary?

lewismc commented on src/plugin/protocol-smb/src/java/org/apache/nutch/protocol/smb/SmbURLConnection.java:
```diff
@@ -0,0 +1,57 @@
+package org.apache.nutch.protocol.smb;
```
Please add ALv2 license header

lewismc reviewed (changes requested):
Hi @HiranChaudhuri I added quite a few comments for your consideration. Thanks for submitting this PR 👍 
Please ping me once your ready and we can go for round # 2 of peer review.

Further out, I think we could implement some testing for this protocol plugin. We could use `testcontainers` and essentially spin up a local Samba server using @nddipiazza 's [smbj-inttest](https://hub.docker.com/r/ndipiazza/smbj-inttest) image. We can come back to this one the PR has evolved a bit.

Commit:
Reduce logging
Improve error handling
Rename class as requested
Added license header
Improve url parsing
added robots.txt

Commit:
fix connection problems

Commit:
better handling of 'invalid network name' (invalid share names)

lewismc commented on src/plugin/protocol-smb/plugin.xml:
```diff
@@ -0,0 +1,53 @@
+   provider-name="Hiran Chaudhuri">
+
+   <runtime>
+      <library name="asn-one-0.6.0.jar"/>
```
If you follow the guidance I gave above regarding adapting the steps in [howto_upgrade_solr.md](https://github.com/apache/nutch/blob/master/src/plugin/indexer-solr/howto_upgrade_solr.md) then this stage will be done for you.

lewismc reviewed (changes requested):
Thanks for all of the responses @HiranChaudhuri please see my updated suggestions.

lewismc commented on src/plugin/protocol-smb/src/java/org/apache/nutch/protocol/smb/Smb.java:
```diff
@@ -0,0 +1,292 @@
+
+import crawlercommons.robots.BaseRobotRules;
+
+public class Smb implements Protocol {
```
Yes I think we can do this in a different ticket.

lewismc commented on src/plugin/protocol-smb/src/java/org/apache/nutch/protocol/smb/Smb.java:
```diff
@@ -0,0 +1,292 @@
+  private Set<String> ignoreFiles;
+
+  public Smb() {
+    // todo: files that should be skipped could be configurable.
```
This TODO is still present so I'm not resolving this conversation yet.

lewismc commented on src/plugin/protocol-smb/src/java/org/apache/nutch/protocol/smb/Smb.java:
```diff
@@ -0,0 +1,292 @@
+  public void setConf(Configuration conf) {
+    this.conf = conf;
+
+    // todo: is it possible to use configuration "per server" or "per share"?
```
You can take a look at [IndexWriters#loadWritersConfiguration(Configuration conf)](https://github.com/apache/nutch/blob/4a61208f492613f2c5282741e64c036acabeb71e/src/java/org/apache/nutch/indexer/IndexWriters.java#L116-L143). Here you can see that we [read an XML file](https://github.com/apache/nutch/blob/4a61208f492613f2c5282741e64c036acabeb71e/conf/index-writers.xml.template) containing several indexing writer plugin configurations.
Maybe we could do something similar for the SMB plugin. This would set the standard for how other protocol plugins could accomplish the same e.g. SMTP, (S)FTP, etc.

lewismc commented on src/plugin/protocol-smb/src/java/org/apache/nutch/protocol/smb/Smb.java:
```diff
@@ -0,0 +1,292 @@
+        LOG.warn("path={}", path);
+
+        // todo: we construct and destruct the connection for each and every URL. Can connection pools improve?
+        SMBClient client = new SMBClient();
```
Yes I think we need to experiment here a bit.

Commit:
project cleanup

Commit:
remove non-required dependencies

lewismc commented on src/plugin/build.xml:
```diff
@@ -78,6 +78,7 @@
     <ant dir="protocol-httpclient" target="deploy"/>
     <ant dir="protocol-interactiveselenium" target="deploy" />
     <ant dir="protocol-okhttp" target="deploy"/>
+    <ant dir="protocol-smb" target="deploy"/>
```
No problem, this is not intuitive and it is easy to miss.

lewismc commented:
Hi @HiranChaudhuri now that you've activated the plugin `test` target, I suggest that we implement the unit tests using the test-containers utility. An example can be found at https://java.testcontainers.org/quickstart/junit_4_quickstart/.
For example we could use 
```
@Rule
public GenericContainer redis = new GenericContainer(DockerImageName.parse("ndipiazza/smbj-inttest:1.1"))
    .withExposedPorts(445);
```
Please let me know your thoughts on this.

Commit:
can use different authentication per URL pattern

lewismc commented on conf/url-authentication.xml.template:
```diff
@@ -0,0 +1,6 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Please provide ALv2 header

lewismc commented on src/plugin/protocol-smb/src/java/org/apache/nutch/protocol/smb/URLAuthentication.java:
```diff
@@ -0,0 +1,120 @@
+package org.apache.nutch.protocol.smb;
```
Please provide ALv2 header

lewismc reviewed (changes requested):
Trivial header requests. Thanks

Commit:
added ALv2 header

Commit:
Cache robots.txt to cut network requests in half

Commit:
bugfix: Instread of trunkating at zero bytes by default grab a maximum of content

lewismc commented on src/plugin/protocol-smb/src/test/org/apache/nutch/protocol/smb/TestSmbProtocol.java:
```diff
@@ -0,0 +1,51 @@
+    @Before
+    public void setUp() {
+        LOG.warn("setUp()");
+        Assert.fail();
```
These tests cannot be committed in the failing state. Thy will destabilize the CI builds. 
Previously I suggested using [test containers](https://issues.apache.org/jira/browse/NUTCH-2856?focusedCommentId=17467111&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-17467111). 
If docker is not available on the host machine when tests are being run then we could use the [@Testcontainers(disabledWithoutDocker = true)](https://java.testcontainers.org/quickstart/junit_5_quickstart/#4-additional-attributes) syntax.

lewismc reviewed (changes requested):
@HiranChaudhuri 
Thanks for coming back to this PR, and thanks for the contribution.
Please see my comments, I am happy to work with you in a bid to get this PR into the Nutch codebase.
